### ISSUE TITLE
Add setting to specify intermediate build path (needed for C++ CLI type library imports)

### ISSUE THREAD

**Issue by kkutsner** [microsoft/vscode-cpptools#2703]
**Type: LanguageService**

- OS and Version: Windows 10
- VS Code Version: 1.28.2
- C/C++ Extension Version: 0.19.0

Intellisense complains about `import "..."` lines in the code and asks to check my include settings.

**To Reproduce**
1. Create new a folder and open it with vscode
2. Create main.cpp file with the following code:
```C++
#include <string>
#import "msxml.dll"
```
3. Open "Problems" window and see the items below, e.g.:
```
#include errors detected. Please update your includePath. IntelliSense features for this translation unit (<>\vscode-bug\main.cpp) will be provided by the Tag Parser.
cannot open source file ".\msxml.tlh"
```
4. If line `#import "msxml.dll"` removed, the "Problems" window become clean.

**Expected behavior**
I believe it should not complain about importing libraries. In the real world case, I tried to specify the include path where TLH file were generated explicitly, but it didn't help.

**Screenshots**
![image](https://user-images.githubusercontent.com/2157905/47322029-660e5600-d60b-11e8-89f6-64835bf1762f.png)

**Comment by sean-mcmanus** [comment]
For C++/CLI we support adding the /clr flag to the end of the compilerPath, but that doesn't fix this issue -- our support for C++/CLI is minimal (i.e. missing #import support). You can set the intelliSenseEngineFallback to "Disabled" to get IntelliSense working in that file, but it won't have symbols from the msxml.dll.

**Comment by Colengms** [comment]
It looks like we need to provide the path that the `.tlh` files are generated to, to our IntelliSense process (using `--import_dir`).  The default value is ".", which is probably always wrong.  VS sets this correctly, as it's also the build system and knows where those files were written (the intermediate output path.  CMake uses a similar intermediate output path).  We would need to add a setting to allow users to specify this path.  We should also add it to the config provider API, so it wouldn't be necessary to set manually when using a config provider.

**Comment by kkutsner** [comment]
I tried to browse the code to see if I can hack it to see if it works, but it wasn't quite obvious.... I could grab the extension and try to build and spend a bunch of time to debug it to understand how everything works, but I wonder if someone can point out were to look at to hard code the `--import_dir` setting to see if it actually works.

**Comment by sean-mcmanus** [comment]
It's possible you might be able to edit the file `"C:\Users\<username>\.vscode\extensions\ms-vscode.cpptools-1.0.1\bin\msvc.json"` or https://github.com/microsoft/vscode-cpptools/blob/master/Extension/bin/msvc.json, e.g.
```
      "-D_USE_DECLSPECS_FOR_SAL=1",
      "--import_dir",
      "<path>"
    ],
```
but I haven't tried it out.

**Comment by kkutsner** [comment]
Thanks @sean-mcmanus! That is actually working! At least it does not show up those errors! Will try to play more with that and will get back with updates.

**Comment by sean-mcmanus** [comment]
@kkutsner Be aware that the downside with that workaround is that edits to msvc.json would stop working and need to be re-applied after an update is released.

**Comment by kkutsner** [comment]
@sean-mcmanus Yeah, it is clear from the fact that msvc.json file is a source code of the extension and not a user config so the consequences are pretty much expected...

**Comment by nmoreaud** [comment]
Update: config file path has changed: 
```
https://github.com/microsoft/vscode-cpptools/blob/master/Extension/bin/windows.msvc.x64.json
C:\Users\XXX\.vscode\extensions\ms-vscode.cpptools-1.18.5-win32-x64\bin\windows.msvc.x64.json
```

Add this to `defaults` array of 
```
      "--import_dir",
      "C:\\dev\\...\\mydir"
```

`D_USE_DECLSPECS_FOR_SAL` is not necessary.

I thought adding this directory to includes list of CMake would be enough, but it's not the case.

**Comment by Colengms** [comment]
If I understand the scenario correctly, I believe this issue (specifically the need for a setting) may already have a solution which doesn't require modifying internal JSON files.  The output (and import) path for generated `.tlh` files is specified using the `/Fo` argument to `cl.exe`.  If configuring the C/C++ extension using a `compilerPath` in `c_cpp_properties.json`, `/Fo` can be specified in the `compilerArgs` field.  If using `compile_commands.json`, the arg should be in the command line(s).  If using a custom configuration provider, it would be responsible for providing this argument and path.

Unfortunately, there is a known issue with CMake Tools, in which it fails to provide some arguments CMake uses to build files, including the `/Fo` argument.  This issue is here: https://github.com/microsoft/vscode-cmake-tools/issues/3085

**Comment by bobbrow** [comment]
Closing this issue as we are not planning to add an additional setting for the scenario.
